src/gui/src/PatternEditor/PatternEditorPanel.cpp

```diff
@@ -884,6 +870,85 @@ void PatternEditorPanel::patternSizeChanged( int nSelected )
 	EventQueue::get_instance()->push_event( EVENT_SELECTED_PATTERN_CHANGED, -1 );
 }
 
+void PatternEditorPanel::updatePatternSizeLCD(){
+	// update pattern size
+	int nPatternSize = m_pPattern->get_length();
+	int den = m_pPattern->get_denominator();
+	int nBeatUnit = MAX_NOTES / den;
+
+	char tmp[20];
+	if (nPatternSize % nBeatUnit == 0) {
+		sprintf( tmp, "%d/%d", nPatternSize / nBeatUnit, den);
+	} 
+	else {
+		sprintf( tmp, "%.3f/%d", (float) nPatternSize / nBeatUnit, den);
+	}
+	__pattern_size_LCD->setText( tmp );//
+}
+
+void PatternEditorPanel::patternSizeLCDClicked()
+{
+	Hydrogen *pEngine = Hydrogen::get_instance();
+	if ( pEngine->getState() != STATE_READY ) {	
+		QMessageBox::information( this, "Hydrogen", tr( "Is not possible to change the pattern size when playing." ) );
+		return;
+	} //TODO is it really impossible to change the pattern size when playing?
+
+	bool bIsOkPressed;
+	int denominator;
+	
+	QString qtmp = QInputDialog::getText( this, "Hydrogen",
+				tr( "New Pattern Size\nIndicate numerator / denominator (eg. \"4/4\") or just numerator"),
+				QLineEdit::Normal, __pattern_size_LCD->getText(), &bIsOkPressed );
+	
+	if ( bIsOkPressed  ) {		
+		QStringList parts = qtmp.split('/');
+		int nDenominator = m_pPattern->get_denominator();
+		if ( parts.size() == 1 || parts.size() == 2) { // must reject if parts.size > 2 or null
+		    bool bOk;
+		    double fNumerator = parts[0].toDouble( &bOk );
+		    if ( bOk && parts.size() == 2 ) {
+			nDenominator = parts[1].toInt( &bOk );
+			if (bOk && (nDenominator <= 0 || nDenominator > 192) ){
+			   QMessageBox::information( this, "Hydrogen", tr( "Denominator value rejected.\nLimits: (0, 192]" ) );
+			   return;
+			}
+		    }
+		    if ( bOk && fNumerator > 0) {
+			if (fNumerator / nDenominator > 4.){ 
+			     //this is limited because the pattern editor ruler goes up to 16/4. Limit might be extended
+				QMessageBox::information( this, "Hydrogen", tr( "Pattern size too big.\nMaximum = 16/4" ) );
+				return;
+			}
+			else {	 
+				if ( MAX_NOTES % nDenominator != 0 ){
+					QMessageBox::information( this, "Hydrogen",
+						tr("Warning: since finite resolution, Hydrogen can handle precisely only "
+						"denominators that divide 192 (i.e. powers of 2, also multiplied by 3)") );
+				}
+				// set numerator and denominator
+				double fLength = MAX_NOTES / (double) nDenominator * fNumerator;
+				if (fLength - round(fLength) != 0){
```

Comment:
I don't know if you can get false-positive hits
I actually thought to enter the long decimals you are sayng. I see that if you enter at least 16 decimal digits (0.6666666666666666 or 0.3333333333333333) the condition is negative. Else is positive

Reply:
What about `fLength % 1 `?

Reply:
`abs(fLength - round(fLength)) < epsilon` ?

Reply:
> `abs(fLength - round(fLength)) > epsilon` ?

That works! I was thinking  of something like `round(fLength * 1000.0)/1000.0` but yours is better!

Reply:
I am still confused, should `epsilon` depend on the LCD precision, say like:

```
double fLength = MAX_NOTES / (double) nDenominator * fNumerator;
double epsilon = 0.001 * MAX_NOTES/nDenominator; //precision of Pattern Size Display
if (abs(fLength - round(fLength)) > epsilon) {
	//Message...
}
fLength = round(fLength); //always round for any case (eg 191.9999 => 192)
m_pPattern->set_length( (int) fLength);
```
or equivalently ` abs(fNumerator - round(fNumerator*1000.)/1000.) > 0.001`

or must `epsilon` depend on the computer calculus approximations?
I don't know if values like` fLength`=191.99999 will ever occur, but I moved the` round()` function out of the scope. The message has the only aim to inform that the LCD display will not show the input value (in its three decimal digits).

Reply:
Sorry, I missed this earlier.

I had been thinking the LCD precision would be appropriate because it seems unreasonable to expect the user to enter more digits than are presented in the display, but on reflection, at least one more digit of precision would be good, to avoid that rounding changing the last visible digit!

Reply:
With 1 more digit there are still some strange cases:
input 1.666/4  => display = 1.667/4, warning message, _message expected_
input 1.6666/4 => display = 1.667/4, NO warning message, _ok since the last digit 7 is the proper approximation_
input 1.6669999/4 => display = 1.667/4, warning message, _but message unexpected since the last digit 7 is the proper approximation, as in the previous case_

I thought  to a different solution.
We can and should always round `fLength ` because it must become a value of type `int`, also this prevents some approximated or incorrect double representation of false non-integers.
Then the following code will warn the user in case the pattern size display won't respect the input value
```
int nLength = round( MAX_NOTES / (double) nDenominator * fNumerator );

// set size and denominator				
m_pPattern->set_length( nLength);
m_pPattern->set_denominator( nDenominator );
patternLengthChanged();

/* Message to tell the user why the pattern display won't respect the input value. 
Note: constant = 1000 since the displayed numerator has 3 decimal digits*/
int displayedNum_x1000 = round( (double) nLength / MAX_NOTES * nDenominator * 1000);
int roundInputNum_x1000 = round( fNumerator * 1000);
if (displayedNum_x1000 != roundInputNum_x1000) {
	QMessageBox::information( this, "Hydrogen",
		tr("Pattern size was rounded.\n(resolution = %1 ticks/quarter note)" ).arg(MAX_NOTES/4));
}
```